plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/RewriteVarcharConstant.java

```diff
@@ -46,6 +48,14 @@ public Optional<String> rewrite(Constant constant, Captures captures, RewriteCon
         if (slice == null) {
             return Optional.empty();
         }
-        return Optional.of("'" + slice.toStringUtf8().replace("'", "''") + "'");
+
+        String sliceUtf8String = slice.toStringUtf8();
+        boolean isUnicodeString = UNICODE_CHARACTER_MATCHER.matchesAnyOf(sliceUtf8String);
+
+        if (isUnicodeString) {
+            return Optional.of("N'" + sliceUtf8String.replace("'", "''") + "'");
+        }
+
+        return Optional.of("'" + sliceUtf8String.replace("'", "''") + "'");
```

Comment:
this shouldn't be in trino-base-jdbc - it's specific to SQL Server.

Reply:
Looks like it's part of  SQL-92 standard and SQL server is required to use it, but all other support it (but not required), like MySQL, Postgres, DB-2 e.t.c, so should we move only for SQL Server part?

Reply:
is the `N` string interpretation same across databases?

Reply:
This what I found about N prefix:
`The "N" prefix stands for National Language in the SQL-92 standard, and is used for representing Unicode characters`
`While most databases do not need the added N prefix when using Unicode data, in SQL Server you must precede all Unicode strings with a prefix N when dealing with Unicode string constants.`

`All other supported backends work with or without the N prefix - the N prefix is not required in those environments but does work if used.`

Reply:
Not all databases / query engines support N prefix. For example, does Trino support it?

Reply:
Supporting it is one thing - but the bigger question is do escapes get handled the same across databases.

From an older offline discussion my conclusion was it might make more sense to make ConnectorExpression pushdown use prepared statements which would make handling this (and other cases) easier.

In case we are able to identify cases where the `N` behaviour differs across databases then IMO we should remove all non-arithmetic pushdowns from this PR and revisit them when we have better mechanisms to support them.

Reply:
So maybe we can move for now this N support as you suggested previously only to SQL Server?

Reply:
@hashhar I tried implement today the path, when we `remove all non-arithmetic pushdowns`
but the problem is, arithmetics push down will work only if we have such rule:

`.add(new RewriteComparison(ImmutableSet.of(RewriteComparison.ComparisonOperator.EQUAL, RewriteComparison.ComparisonOperator.NOT_EQUAL)))`

But if add such rule, equal/not equal push down start work even for simple cases with varchars like failing one:
`WHERE variable = 'łąka for the win'`

So look like we need this UnicodeConstantRewrite rule, if we want arithmetics pushdown, but we can keep everything only in SQL Server

Reply:
Let's keep it only in SQL Server for now in that case until we have better ways to implement connector expression to SQL conversion + more test coverage.

The problem seems to be that there is no way to "prevent a connector expression pushdown" - we can only add more rules. i.e. we cannot say allow RewriteComparision but only if operands are integer types (we can but it'd require writing entirely new rule).

Reply:
> The problem seems to be that there is no way to "prevent a connector expression pushdown" - we can only add more rules. i.e. we cannot say allow RewriteComparision but only if operands are integer types (we can but it'd require writing entirely new rule).

exactly, I tried to made it, but with no luck